scripts/algorithms/knn.dml

```diff
@@ -0,0 +1,745 @@
+        if( i_cl_type == 2 ){
+            t_cl_value = as.scalar( rowIndexMax( t( table( NNR_tmp_matrix,1 ) ) ) )
+            #if( PRPath == " " ){
+            #    print( "cate is : " + t_cl_value );
+            #}
+        }else{
+            if ( predict_con_tg == 0){
+                t_cl_value = mean( NNR_tmp_matrix );
+            } else if(predict_con_tg == 1){
+                t_cl_value = median( NNR_tmp_matrix );
+            }
+            #if( PRPath == " " ){
+            #    print( "value is : " + t_cl_value );
+            #}
+        }
+        CL_matrix[i,1] = t_cl_value;
+    }
+}
+if( PRPath != " " ){
+    write( CL_matrix,PRPath,format=FMT );
+}
+
+#naive knn search implement
+naiveKNNsearch = function(
+    matrix[double] P,
+    matrix[double] Q,
+    integer K
+)return(
+    matrix[double] O
+){
+    num_records = nrow (P);
+    num_features = ncol (P);
+    num_queries = nrow (Q);
+    Qt = t(Q);
+    PQt = P %*% Qt;
+    P2 = rowSums (P ^ 2);
+    D = -2 * PQt + P2;
+    if (K == 1) {
+        Dt = t(D);
+        O = rowIndexMin (Dt);
+    } else {
+        O = matrix (0, rows = num_queries, cols = K);
+        parfor (i in 1:num_queries) {
+            D_sorted=order(target=D[,i], by=1, decreasing=FALSE, index.return=TRUE);
+            O[i,] = t(D_sorted[1:K,1]);
+        }   
+    }   
+}
+
+# prepareKNNData:
+#   Do data prepare - dummy for category variable and [-1,1] transform for continues variable 
+# Argument:
+# in_m_data                     input matrix as features
+# in_m_data_type                input matrix with 1 row denoting whether columns in X are 
+#                               continuous ( =1 ) or categorical ( =2 )
+# in_i_trans_continues          Option flag for continuous feature transformed:
+#                               0 = do not transform continuous variable;
+#                               1 = transform continuous variable;
+prepareKNNData = function ( 
+    matrix[double] in_m_data,
+    matrix[double] in_m_data_type,
+    integer in_i_trans_continues
+     )
+return( 
+    matrix[double] out_m_data
+     )
+{
+    i_outmatrix_col = 0;
+    for( i in 1:ncol( in_m_data ) ){
+        #categorical
+        if( as.scalar( in_m_data_type[1,i] ) == 2 ){
+            i_outmatrix_col = i_outmatrix_col + as.integer( max( in_m_data[,i] ) );
+        }else{ #continuous
+            i_outmatrix_col = i_outmatrix_col + 1;
+        }
+    }
+
+    out_m_data = matrix( 0,nrow( in_m_data ),i_outmatrix_col );
+    i_outmatrix_iter = 1; # column iter for transform
+    for( i in 1:ncol( in_m_data ) ){
+        #categorical
+        if( as.scalar( in_m_data_type[1,i] ) == 2 ){
+            tmp_i_max = max( in_m_data[,i] );
+            for( i_c in 1:tmp_i_max ){
+                out_m_data[,i_outmatrix_iter + i_c - 1] = ppred( in_m_data[,i],i_c,"==" );
+            }
+            #iter
+            i_outmatrix_iter = i_outmatrix_iter + as.integer( max( in_m_data[,i] ) );
+        }else{ #continuous
```

Comment:
rescaling continuous variables can be done one the whole input matrix without going through each column in the for loop.